financial-templates-lib/UniswapPriceFeed.js

```diff
@@ -34,8 +40,26 @@ class UniswapPriceFeed {
       return a.logIndex - b.logIndex;
     });
 
+    // Search backwards through the array and grab block timestamps for everything in our lookback window.
+    const currentTime = this.getTime();
+    let i = events.length;
+    while (i !== 0) {
```

Comment:
I'm having a hard time convincing myself that this loop is correct, but the passing test cases indicate that it must be. E.g., if all events are in the window, we get down to `i==0`, and then we index `[--i]`. Shouldn't this operation be out of bounds?

Do you think a normal for-loop could work here?

Reply:
This is because the decrement happens after the while check, but before being used in the array access (pre-decrement). So on the last iteration, `i` is 1 in the while check, then it's decremented to 0 before being used to access `events`. When it loops back around, the while check causes it to break. It's definitely a confusing loop structure.

I originally had it as a decrementing for loop (a little confusing in its own right), but this created an inconsistency in the break case and the non break case. In the break case, `i` would finish as the last index that we want to keep. In the non-break case, we would end up with `-1`. This while loop was to avoid a strange `if` statement after the loop. Happy to move to that or something else if you think that'd be easier to read. For reference, here's that implementation (you could set the if statement up in a few different ways):
```
    const currentTime = this.getTime();
    let i;
    for (i = events.length - 1; i >= 0; i--) {
      const event = events[i];
      event.timestamp = (await web3.eth.getBlock(event.blockNumber)).timestamp;
      event.price = this._getPriceFromSyncEvent(event);
      if (event.timestamp <= currentTime - this.twapLength) {
        break;
      }
    }

    // Cut off all the events that were before the time we care about.
    if (i === -1) {
      this.events = events;
    } else {
      this.events = events.slice(i);
    }
```

Reply:
I guess one way to avoid the explicit `if` in the for loop impl is to do something like:
```
this.events = events.slice(Math.max(i,0));
```

WDYT?

Reply:
The way you have it now is fine too.

Reply:
I feel like the for loop is actually slightly more confusing, so I'll leave as is, but definitely open to changing if you disagree.

Reply:
I feel like the for loop is actually slightly more confusing, so I'll leave as is, but definitely open to changing if you disagree.